### PR TITLE
Remove exit() call from shared libraries

### PR DESCRIPTION
**Pull request by minghui-liu**
#### Bug/feature Description
rpmlint gave a warning `W: shared-lib-calls-exit`. It is unhappy about using `exit()` in a library.

#### Affected Platform(s)
RHEL, CentOS, openSUSE

#### Solution Description
* Replaced some `exit()` calls with `return` statement
* Changed some function's return value from `void` to `int` and let it return an error code. Then the caller calls `exit()`. Essentially this is moving exit() calls from the library code to the calling application code.

#### Testing logs/output
* [Build and test result.txt](https://github.com/PBSPro/pbspro/files/2567827/Build.and.test.result.txt)

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: src/resmom/mom_main.c [hunk 1]
```diff
@@ -1306,7 +1306,8 @@ initialize(void)
 	 *	Check that there are no bad combinations of sharing values
 	 *	across the vnodes.
 	 */
-	avl_create_index(&ix, AVL_NO_DUP_KEYS, 0);
+	if (avl_create_index(&ix, AVL_NO_DUP_KEYS, 0))
+		exit(1);
```
Comment (bayucan) [inline src/resmom/mom_main.c]: I think you need to do a die(0) here so as to cleanup stuff,.. like in a later part of the code.

File: src/cmds/cmds_common.c [hunk 2]
```diff
@@ -35,35 +35,52 @@
...
+ * @param[in]     attrib_value - attribute value
  *
- * @param[in] argc - num of args
- * @param[in] argv - pointer args list
+ * @return	Void
+ */
+void
+set_attr(struct attrl **attrib, char *attrib_name, char *attrib_value) {
+    if (_set_attr(attrib, attrib_name, attrib_value))
+        exit(2);
```
[the remark this answers is not in the document]
Comment (mike0042) [reply src/cmds/cmds_common.c]: Oops... guess not. Thought this was in libifl, but it's not. Please disregard.

File: src/cmds/cmds_common.c [hunk 3]
```diff
@@ -35,35 +35,52 @@
...
-execution_mode(int argc, char** argv)
+set_attr_resc(struct attrl **attrib, char *attrib_name, char *attrib_resc, char *attrib_value)
 {
-
-	if (argc == 2 && strcasecmp(argv[1], "--version") == 0) {
-		fprintf(stdout, "pbs_version = %s\n", pbs_version);
-		exit(0);
-	}
+    if (_set_attr_resc(attrib, attrib_name, attrib_resc, attrib_value))
+        exit(2);
```
[the remark this answers is not in the document]
Comment (mike0042) [reply src/cmds/cmds_common.c]: Oops... guess not. Thought this was in libifl, but it's not. Please disregard.

File: src/cmds/qsub.c [hunk 4]
```diff
@@ -2115,7 +2115,9 @@ set_sig_handlers(void)
 	signal(SIGBREAK, win_blockint);
 	signal(SIGTERM, win_blockint);
 
-	winsock_init();
+	if (winsock_init())
+		exit_qsub(1);
+
```
[the remark this answers is not in the document]
Comment (mike0042) [reply src/cmds/qsub.c]: Thanks for addressing.

File: src/tools/tracejob.c [hunk 5]
```diff
@@ -150,7 +150,10 @@ main(int argc, char *argv[])
 	int unknw_job = 0;
 
 	/*the real deal or output pbs_version and exit?*/
-	execution_mode(argc, argv);
+	if (argc == 2 && strcasecmp(argv[1], "--version") == 0) {
```
[the remark this answers is not in the document]
Comment (mike0042) [reply src/tools/tracejob.c]: After your latest change you still need to add parenthesis around (strcasecmp(argv[1], "--version") == 0) in the conditional. Another advantage to making this a macro is that you only need to update it in one place.

**Comment by mike0042** [comment]
Looks like the AppVeyor build failed...
```
mom_hook_func.c
c:\projects\pbspro\src\include\pbs_version.h(64) : error C2059: syntax error : 'if'
c:\projects\pbspro\src\include\pbs_version.h(64) : error C2017: illegal escape sequence
c:\projects\pbspro\src\include\pbs_version.h(65) : error C2017: illegal escape sequence
c:\projects\pbspro\src\include\pbs_version.h(66) : error C2017: illegal escape sequence
c:\projects\pbspro\src\include\pbs_version.h(67) : error C2017: illegal escape sequence
c:\projects\pbspro\src\include\pbs_version.h(68) : error C2059: syntax error : '}'
c:\projects\pbspro\src\include\pbs_version.h(68) : error C2059: syntax error : ')'
..\..\src\resmom\mom_hook_func.c(202) : error C2065: 'run_exit' : undeclared identifier
153>..\..\src\resmom\mom_hook_func.c(231) : warning C4018: '<' : signed/unsigned mismatch
153>..\..\src\resmom\mom_hook_func.c(236) : warning C4018: '<' : signed/unsigned mismatch
..\..\src\resmom\mom_hook_func.c(785) : error C2065: 'run_exit' : undeclared identifier
..\..\src\resmom\mom_hook_func.c(840) : error C2065: 'run_exit' : undeclared identifier
..\..\src\resmom\mom_hook_func.c(1412) : error C2065: 'run_exit' : undeclared identifier
..\..\src\resmom\mom_hook_func.c(1416) : error C2065: 'run_exit' : undeclared identifier
..\..\src\resmom\mom_hook_func.c(1430) : error C2065: 'run_exit' : undeclared identifier
..\..\src\resmom\mom_hook_func.c(1431) : error C2065: 'run_exit' : undeclared identifier
..\..\src\resmom\mom_hook_func.c(1433) : error C2065: 'run_exit' : undeclared identifier
..\..\src\resmom\mom_hook_func.c(1448) : error C2065: 'run_exit' : undeclared identifier
..\..\src\resmom\mom_hook_func.c(1449) : error C2065: 'run_exit' : undeclared identifier
..\..\src\resmom\mom_hook_func.c(1604) : error C2065: 'run_exit' : undeclared identifier
```

File: src/lib/Liblog/pbs_log.c [hunk 6]
```diff
@@ -607,9 +607,9 @@ log_open_main(char *filename, char *directory, int silent)
 
 		if (!silent) {
 			log_record(PBSEVENT_SYSTEM, PBS_EVENTCLASS_SERVER, LOG_INFO, "Log", "Log opened");
-			snprintf(tbuf, LOG_BUF_SIZE, "pbs_version=%s", pbs_version);
+			snprintf(tbuf, LOG_BUF_SIZE, "pbs_version=%s", PBS_VERSION);
 			log_record(PBSEVENT_SYSTEM, PBS_EVENTCLASS_SERVER, LOG_INFO, msg_daemonname, tbuf);
-			snprintf(tbuf, LOG_BUF_SIZE, "pbs_build=%s", pbs_build);
+			snprintf(tbuf, LOG_BUF_SIZE, "pbs_build=%s", PBS_BUILD);
```
[the remark this answers is not in the document]
Comment (mike0042) [reply src/lib/Liblog/pbs_log.c]: It is not necessary for this PR.

**Comment by mike0042** [comment]
The script win_configure/projects.VS2008/gen_pbs_version.bat generates pbs_version.h from pbs_version.h.in. I wonder if it is somehow stripping off the backslash characters or adding whitespace after them. We would need to get ahold of the build artifacts in order to tell. I wrote a small program to test this in VS2017 and it works properly. Not sure if it behaves differently than VS2008.

**Comment by minghui-liu** [comment]
Rebased and squashed.

**Pull request merged**